Elektra Web 1.2

omnidan opened the pull request:
changes from #1759

**ToDo**
- [x] visibility
- [x] performance improvements
- [x] rename clusterd
- [x] improve creating instance
- [x] improve single instance mode

-> see https://github.com/ElektraInitiative/libelektra/issues/1845

markus2330 commented on src/tools/web/client/src/components/TreeItem/dialogs/AdditionalSubDialog.jsx:
```diff
@@ -0,0 +1,132 @@
+/**
+ * @file
+ *
+ * @brief sub-dialog to add and modify additional metadata of keys
```
Name of file is too generic.

markus2330 commented on src/tools/web/client/src/components/TreeItem/dialogs/AdditionalSubDialog.jsx:
```diff
@@ -0,0 +1,132 @@
+    }
+  }
+
+  componentWillReceiveProps (nextProps) {
```
Please add some comments (e.g. here).

markus2330 commented on src/tools/web/client/src/reducers/kdb.js:
```diff
@@ -38,6 +38,13 @@ export default function keyReducer (state = {}, action) {
     case SET_KEY_REQUEST:
       return updateState(state, action.request)
 
+    // TODO: recursively copy keys here? is this needed? we refresh on expand anyway
```
todo

omnidan replied:
this isn't a problem anymore as I'm handling the refreshing differently now

markus2330 commented on src/tools/web/client/src/utils.js:
```diff
@@ -0,0 +1,23 @@
+    case 'unsigned_long':
+    case 'long_long':
+    case 'unsigned_long_long':
+    case 'any':
```
'any' is not necessarily a number?

markus2330 commented:
CMake integration is still missing. Please also check for necessary nodejs version.

I thought you'll rename clusterd? The name is not suitable anymore?

I cannot run clusterd with nodejs version: 6.11.4-1 on Debian stretch.

```
npm start

> @elektra-web/clusterd@1.1.0 start /home/markus/Projekte/Elektra/libelektra/src/tools/web/clusterd
> npm run build && npm run start:prod


> @elektra-web/clusterd@1.1.0 build /home/markus/Projekte/Elektra/libelektra/src/tools/web/clusterd
> cross-env NODE_ENV=production babel -d build/ src/ && npm run build --prefix ../client

src/app.js -> build/app.js
src/config.js -> build/config.js
src/connector.js -> build/connector.js
src/db/clusters.js -> build/db/clusters.js
src/db/config.js -> build/db/config.js
src/db/index.js -> build/db/index.js
src/db/instances.js -> build/db/instances.js
src/db/utils.js -> build/db/utils.js
src/index.js -> build/index.js
src/log.js -> build/log.js
src/routes/clusters.js -> build/routes/clusters.js
src/routes/index.js -> build/routes/index.js
src/routes/instances.js -> build/routes/instances.js
src/routes/utils.js -> build/routes/utils.js
src/serveClient.js -> build/serveClient.js
src/versions.js -> build/versions.js

> @elektra-web/client@1.1.0 build /home/markus/Projekte/Elektra/libelektra/src/tools/web/client
> react-scripts build

sh: 1: react-scripts: not found

npm ERR! Linux 4.9.0-6-amd64
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "run" "build" "--prefix" "../client"
npm ERR! node v6.11.4
npm ERR! npm  v3.10.10
npm ERR! file sh
npm ERR! code ELIFECYCLE
npm ERR! errno ENOENT
npm ERR! syscall spawn
npm ERR! @elektra-web/client@1.1.0 build: `react-scripts build`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the @elektra-web/client@1.1.0 build script 'react-scripts build'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the @elektra-web/client package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     react-scripts build
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs @elektra-web/client
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls @elektra-web/client
npm ERR! There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?
```

with following additional information:

```
0 info it worked if it ends with ok
1 verbose cli [ '/usr/bin/node', '/usr/bin/npm', 'start' ]
2 info using npm@3.10.10
3 info using node@v6.11.4
4 verbose run-script [ 'prestart', 'start', 'poststart' ]
5 info lifecycle @elektra-web/clusterd@1.1.0~prestart: @elektra-web/clusterd@1.1.0
6 silly lifecycle @elektra-web/clusterd@1.1.0~prestart: no script for prestart, continuing
7 info lifecycle @elektra-web/clusterd@1.1.0~start: @elektra-web/clusterd@1.1.0
8 verbose lifecycle @elektra-web/clusterd@1.1.0~start: unsafe-perm in lifecycle true
9 verbose lifecycle @elektra-web/clusterd@1.1.0~start: PATH: /usr/lib/node_modules/npm/bin/node-gyp-bin:/home/markus/Projekte/Elektra/libelektra/src/tools/web/clusterd/node_modules/.bin:/usr/lib/git-core:/usr/local/bin:/usr/games:/usr/bin:/bin:/home/markus/bin:/opt/idea/bin:.
10 verbose lifecycle @elektra-web/clusterd@1.1.0~start: CWD: /home/markus/Projekte/Elektra/libelektra/src/tools/web/clusterd
11 silly lifecycle @elektra-web/clusterd@1.1.0~start: Args: [ '-c', 'npm run build && npm run start:prod' ]
12 silly lifecycle @elektra-web/clusterd@1.1.0~start: Returned: code: 1  signal: null
13 info lifecycle @elektra-web/clusterd@1.1.0~start: Failed to exec start script
14 verbose stack Error: @elektra-web/clusterd@1.1.0 start: `npm run build && npm run start:prod`
14 verbose stack Exit status 1
14 verbose stack     at EventEmitter.<anonymous> (/usr/lib/node_modules/npm/lib/utils/lifecycle.js:255:16)
14 verbose stack     at emitTwo (events.js:106:13)
14 verbose stack     at EventEmitter.emit (events.js:191:7)
14 verbose stack     at ChildProcess.<anonymous> (/usr/lib/node_modules/npm/lib/utils/spawn.js:40:14)
14 verbose stack     at emitTwo (events.js:106:13)
14 verbose stack     at ChildProcess.emit (events.js:191:7)
14 verbose stack     at maybeClose (internal/child_process.js:920:16)
14 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:230:5)
15 verbose pkgid @elektra-web/clusterd@1.1.0
16 verbose cwd /home/markus/Projekte/Elektra/libelektra/src/tools/web/clusterd
17 error Linux 4.9.0-6-amd64
18 error argv "/usr/bin/node" "/usr/bin/npm" "start"
19 error node v6.11.4
20 error npm  v3.10.10
21 error code ELIFECYCLE
22 error @elektra-web/clusterd@1.1.0 start: `npm run build && npm run start:prod`
22 error Exit status 1
23 error Failed at the @elektra-web/clusterd@1.1.0 start script 'npm run build && npm run start:prod'.
23 error Make sure you have the latest version of node.js and npm installed.
23 error If you do, this is most likely a problem with the @elektra-web/clusterd package,
23 error not with npm itself.
23 error Tell the author that this fails on your system:
23 error     npm run build && npm run start:prod
23 error You can get information on how to open an issue for this project with:
23 error     npm bugs @elektra-web/clusterd
23 error Or if that isn't available, you can get their info via:
23 error     npm owner ls @elektra-web/clusterd
23 error There is likely additional logging output above.
24 verbose exit [ 1, true ]
```

Please also add enum+visiblity support.

omnidan commented:
@markus2330 I renamed clusterd to webd.

run `npm install` in the client directory. you now have to `npm install` and `npm start` the client instead of clusterd (see updated README.md).

I updated the package scripts so that `npm install` in the `webd` directory will also install client dependencies. now you can install & start either webd or the client, it will always start both as they are always used together (webd serves the client).

I use node 9.3.0. I also tested it with 6.11.4 and everything works fine.

I also improved the performance a lot by grouping requests together. I tested it with a keyset of 100 keys, with 10 subkeys each and it loads relatively fast (the server does take ~1 second to finish loading the 100 keys, the bottleneck here being executing `kdb get` instead of the api directly).

omnidan commented:
@markus2330 regarding the cmake integration - what should the cmake integration do exactly? do you have a template config I could use? I have never configured cmake in a project before.

markus2330 commented:
> see updated README.md

There seems to be inconsistencies between the various READMEs. In the main README.md it says little about webd, please update what the differences between webd and client are.

> run npm install in the client directory.

After various errors which could be resolved by repeating "npm install", I could now connect.

Why is it called client not clientd?

> regarding the cmake integration - what should the cmake integration do exactly?

Install necessary files and call npm like done for src/tools/rest-frontend. Not a high-priority task, though. Better if you today implement the remaining feature, test them, and improve the usability.

markus2330 commented:
> unfortunately adding visibility is not too easy because I would have to filter data on an upper layer where metakeys are not available yet (for performance reasons)

Why not hide it after the fact? It is not a security feature, it is only about usability.

omnidan commented:
@markus2330 all tasks from #1845 are done now

markus2330 commented:
Great!

I am missing a README which explains the structure of the code, e.g., how to add support for new meta data and similar.

And please rebase, the PR crosses the release tag. It would be great (but it is not necessary) if the commit message would always start with `web-ui:` or similar. (But please keep the commits, its great to have them.)

As said I will look at it in more details asap.

markus2330 commented:
Thank you for the improvements.

Unfortunately there are still many grave usability problems, see #1853

markus2330 commented:
And please update doc/METADATA.ini with `usedby/tool = web` for all properties that you support.

omnidan commented:
@markus2330 I fixed all the issues I could reproduce from #1853 and the left over issues from #1845. I also rebased the PR, added documentation about the code structure in the README.md, fixed the cmake config (also documented in README.md now) and updated doc/METADATA.ini.

markus2330 commented:
I was unable to run webd :cry: 

See #1864

Please fix the npm deps and the cmake integration first so that we can merge this PR (please set "ready to merge", further fixes in a 1.3 version. This PR and its accompanied issues are becoming confusing.). Then I'll recheck if the usability problems are gone so that we can start the usability study soon.

markus2330 commented:
Starting elektrad fails:

```

> @elektra-web/elektrad@1.1.0 start /home/markus/Projekte/Elektra/libelektra/src/tools/web/elektrad
> npm run build && npm run start:prod


> @elektra-web/elektrad@1.1.0 build /home/markus/Projekte/Elektra/libelektra/src/tools/web/elektrad
> cross-env NODE_ENV=production babel -d build/ src/

sh: 1: cross-env: not found

npm ERR! Linux 4.12.0-0.bpo.2-amd64
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "run" "build"
npm ERR! node v6.13.1
npm ERR! npm  v3.10.10
npm ERR! file sh
npm ERR! code ELIFECYCLE
npm ERR! errno ENOENT
npm ERR! syscall spawn
npm ERR! @elektra-web/elektrad@1.1.0 build: `cross-env NODE_ENV=production babel -d build/ src/`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the @elektra-web/elektrad@1.1.0 build script 'cross-env NODE_ENV=production babel -d build/ src/'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the @elektra-web/elektrad package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     cross-env NODE_ENV=production babel -d build/ src/
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs @elektra-web/elektrad
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls @elektra-web/elektrad
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/markus/Projekte/Elektra/libelektra/src/tools/web/elektrad/npm-debug.log

npm ERR! Linux 4.12.0-0.bpo.2-amd64
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "start"
npm ERR! node v6.13.1
npm ERR! npm  v3.10.10
npm ERR! code ELIFECYCLE
npm ERR! @elektra-web/elektrad@1.1.0 start: `npm run build && npm run start:prod`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @elektra-web/elektrad@1.1.0 start script 'npm run build && npm run start:prod'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the @elektra-web/elektrad package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     npm run build && npm run start:prod
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs @elektra-web/elektrad
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls @elektra-web/elektrad
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/markus/Projekte/Elektra/libelektra/src/tools/web/elektrad/npm-debug.log
```

```
0 info it worked if it ends with ok
1 verbose cli [ '/usr/bin/node', '/usr/bin/npm', 'start' ]
2 info using npm@3.10.10
3 info using node@v6.13.1
4 verbose run-script [ 'prestart', 'start', 'poststart' ]
5 info lifecycle @elektra-web/elektrad@1.1.0~prestart: @elektra-web/elektrad@1.1.0
6 silly lifecycle @elektra-web/elektrad@1.1.0~prestart: no script for prestart, continuing
7 info lifecycle @elektra-web/elektrad@1.1.0~start: @elektra-web/elektrad@1.1.0
8 verbose lifecycle @elektra-web/elektrad@1.1.0~start: unsafe-perm in lifecycle true
9 verbose lifecycle @elektra-web/elektrad@1.1.0~start: PATH: /usr/lib/node_modules/npm/bin/node-gyp-bin:/home/markus/Projekte/Elektra/libelektra/src/tools/web/elektrad/node_modules/.bin:/usr/lib/git-core:/usr/local/bin:/usr/games:/usr/bin:/bin:/home/markus/bin:/opt/idea/bin:.
10 verbose lifecycle @elektra-web/elektrad@1.1.0~start: CWD: /home/markus/Projekte/Elektra/libelektra/src/tools/web/elektrad
11 silly lifecycle @elektra-web/elektrad@1.1.0~start: Args: [ '-c', 'npm run build && npm run start:prod' ]
12 silly lifecycle @elektra-web/elektrad@1.1.0~start: Returned: code: 1  signal: null
13 info lifecycle @elektra-web/elektrad@1.1.0~start: Failed to exec start script
14 verbose stack Error: @elektra-web/elektrad@1.1.0 start: `npm run build && npm run start:prod`
14 verbose stack Exit status 1
14 verbose stack     at EventEmitter.<anonymous> (/usr/lib/node_modules/npm/lib/utils/lifecycle.js:255:16)
14 verbose stack     at emitTwo (events.js:106:13)
14 verbose stack     at EventEmitter.emit (events.js:191:7)
14 verbose stack     at ChildProcess.<anonymous> (/usr/lib/node_modules/npm/lib/utils/spawn.js:40:14)
14 verbose stack     at emitTwo (events.js:106:13)
14 verbose stack     at ChildProcess.emit (events.js:191:7)
14 verbose stack     at maybeClose (internal/child_process.js:920:16)
14 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:230:5)
15 verbose pkgid @elektra-web/elektrad@1.1.0
16 verbose cwd /home/markus/Projekte/Elektra/libelektra/src/tools/web/elektrad
17 error Linux 4.12.0-0.bpo.2-amd64
18 error argv "/usr/bin/node" "/usr/bin/npm" "start"
19 error node v6.13.1
20 error npm  v3.10.10
21 error code ELIFECYCLE
22 error @elektra-web/elektrad@1.1.0 start: `npm run build && npm run start:prod`
22 error Exit status 1
23 error Failed at the @elektra-web/elektrad@1.1.0 start script 'npm run build && npm run start:prod'.
23 error Make sure you have the latest version of node.js and npm installed.
23 error If you do, this is most likely a problem with the @elektra-web/elektrad package,
23 error not with npm itself.
23 error Tell the author that this fails on your system:
23 error     npm run build && npm run start:prod
23 error You can get information on how to open an issue for this project with:
23 error     npm bugs @elektra-web/elektrad
23 error Or if that isn't available, you can get their info via:
23 error     npm owner ls @elektra-web/elektrad
23 error There is likely additional logging output above.
24 verbose exit [ 1, true ]
```

omnidan commented:
@markus2330 did you follow the tasks from README.md?

The following works fine for me with a fresh clone and clean installation:

- `cmake .. -DTOOLS="web"`
- `make -j8`
- `sudo make install`
- `kdb run-elektrad`
- `kdb run-web`

It seems like you started it from source and did not run `npm install` before `npm start` (also documented in README.md)?

markus2330 commented:
> The following works fine for me with a fresh clone and clean installation:

I think the problem is rather somewhere in the npm cache/history. I just retried on another computer and it worked there.

> The following also works for me with a fresh clone (also documented in README.md):

I always did it like written in the README.md. Install had no problems (except of these "UNMET PEER DEPENDENCY" written in red all-upper case.)

> The following works fine for me with a fresh clone and clean installation:

Yes, works for me but with some warnings:

```
...
-- Installing: /usr/local/elektra/lib/elektra/tool_exec/build-web
-- Installing: elektra-web > elektrad
npm WARN deprecated babel-preset-es2015@6.24.1: 🙌
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@^1.0.0 (node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.1.3: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm WARN @elektra-web/elektrad@1.1.0 No repository field.
npm WARN @elektra-web/elektrad@1.1.0 license should be a valid SPDX license expression
-- Installing: elektra-web > webd
npm WARN deprecated babel-preset-es2015@6.24.1: 🙌
npm WARN deprecated node-uuid@1.4.8: Use uuid module instead
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@^1.0.0 (node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.1.3: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm WARN debug-dude@1.0.3 requires a peer of debug@^2.2.0 but none was installed.
npm WARN babel-loader@7.1.4 requires a peer of webpack@2 || 3 || 4 but none was installed.
npm WARN eslint-config-standard@10.2.1 requires a peer of eslint-plugin-import@>=2.2.0 but none was installed.
npm WARN eslint-config-standard@10.2.1 requires a peer of eslint-plugin-node@>=4.2.2 but none was installed.
npm WARN @elektra-web/webd@1.1.0 No repository field.
npm WARN @elektra-web/webd@1.1.0 license should be a valid SPDX license expression
-- Installing: elektra-web > client
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@^1.0.0 (node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.1.3: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm WARN babel-loader@7.1.4 requires a peer of webpack@2 || 3 || 4 but none was installed.
npm WARN debug-dude@1.0.3 requires a peer of debug@^2.2.0 but none was installed.
npm WARN eslint-config-standard@10.2.1 requires a peer of eslint-plugin-import@>=2.2.0 but none was installed.
npm WARN eslint-config-standard@10.2.1 requires a peer of eslint-plugin-node@>=4.2.2 but none was installed.
npm WARN @elektra-web/webd@1.1.0 No repository field.
npm WARN @elektra-web/webd@1.1.0 license should be a valid SPDX license expression
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.1.2 (node_modules/react-scripts/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.1.2: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@^1.1.2 (node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.1.3: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@^1.0.0 (node_modules/webpack-dev-server/node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.1.3: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm WARN bosket-react@0.4.5 requires a peer of @bosket/core@^0.4.5 but none was installed.
npm WARN bosket-react@0.4.5 requires a peer of @bosket/tools@^0.4.5 but none was installed.
npm WARN ajv-keywords@3.1.0 requires a peer of ajv@^6.0.0 but none was installed.
npm WARN @elektra-web/client@1.1.0 No repository field.
npm WARN @elektra-web/client@1.1.0 license should be a valid SPDX license expression
-- Installing: /usr/local/elektra/share/man/man1/kdb-run-elektrad.1
...
```

The installation is huge, though. It is > 500MB.

Let us debug the problems of the other machine separately.

markus2330 commented:
Is it ready to merge now?

omnidan commented:
@markus2330 that is quite weird. Yes, it is ready to merge now.

> Let us debug the problems of the other machine separately.

I can have a look at it at the next meeting.

omnidan commented:
> The installation is huge, though. It is > 500MB.

Welcome to the wonderful world of JavaScript :grin: Most of these are build-dependencies (devDependencies) though, the actual size after building including runtime dependencies is probably much smaller.

markus2330 commented:
Thank you! I am currently moving all remaining issues to #1865

markus2330 commented:
I think I found the root cause of the many troubles I have (and you don't). My system/elektra seems to cause an endless recursion: #1867

This needs to be fixed first because many other problems might be directly or indirectly affected by this.
